---
title: "Class 15 Gene Ontology"
output: github_document
---

# Section 1: Differential Expression Analysis 

```{r}

# load this package we need 

library(DESeq2)

```

First we read in the files 

```{r }
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

```


Import metadata and view it 

```{r }

colData = read.csv(metaFile, row.names=1)
head(colData)


```

Import countdata calculated through RNA-seq

```{r }

countData = read.csv(countFile, row.names=1)

#view 
head(countData)

```


Note we need to remove the odd first length column. Then we can merge our two files by sample name


```{r}

countData <- as.matrix(countData[,2:7])

#check to make sure we removed the right thing

head(countData)

```

In our data there are many genes where there is 0 expression across all of the samples. We want to remove these for downstream analysis. 


```{r}

countData[countData==0] <- NA
countData = na.omit(countData)

#check that it worked 
head(countData)

```

# Running DESeq2 on our count data

```{r}

# feed in our countData and colData 

dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
dds = DESeq(dds)

```

```{r}

dds

```

```{r}

# view our results 

res = results(dds)

```

View summary of how many genes are up or downregulated 

```{r}

summary(res)

```

# Plotting a volcano plot 

Traditional format for displaying differential expression data. We plot log2FoldChange of gene expression against adjusted p value. 

```{r}

plot( res$log2FoldChange, -log(res$padj) )

```


This is hard to read so we are going to add color to highlight the differentially expressed genes that we are most interested in. 

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res) )

# Color red the genes with absolute fold change above 2
mycols[ abs(res$log2FoldChange) > 2 ] <- "red"

# Color blue those with adjusted p-value less than 0.01
#  and absolute fold change more than 2
inds <- (res$padj < .01) & (abs(res$log2FoldChange) > 2 )
mycols[ inds ] <- "blue"

plot( res$log2FoldChange, -log(res$padj), col=mycols, xlab="Log2(FoldChange)", ylab="-Log(P-value)" )
```


# Adding Gene Annotations

```{r}

#load these packages:

library("AnnotationDbi")
library("org.Hs.eg.db")

```


For this part we are going to change the gene annotation from Ensembl to Entrez

```{r}

# list all of the annotations available
columns(org.Hs.eg.db)

```

```{r}

# Add symbol annotations 
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(res), 
                    keytype="ENSEMBL",
                    column="SYMBOL",
                    multiVals="first")

# Add ENTREZ annotations 
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")

# Add Genename annotations
res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")

# view the res file with our new added annotations 
head(res, 10)

```

```{r}

# Order the results by their adjusted p value 
res = res[order(res$pvalue),]

# Write the results of our analysis to a csv 
write.csv(res, file="deseq_results.csv")

```

# Section 2: Pathway Analysis

We are going to use packages to analyse if the genes that are up and down are enriched in certain biological pathways. 

Background on KEGG pathways:
The gageData package has pre-compiled databases mapping genes to KEGG pathways and GO terms for common organisms. kegg.sets.hs is a named list of 229 elements. Each element is a character vector of member gene Entrez IDs for a single KEGG pathway. (See also go.sets.hs). The sigmet.idx.hs is an index of numbers of signaling and metabolic pathways in kegg.set.gs. In other words, KEGG pathway include other types of pathway definitions, like "Global Map" and "Human Diseases", which may be undesirable in a particular pathway analysis. Therefore,  kegg.sets.hs[sigmet.idx.hs] gives you the "cleaner" gene sets of signaling and metabolic pathways only.



Install these relevant packages in the console:
BiocManager::install( c("pathview", "gage", "gageData") )

```{r}

# load all of the packages we will need 

library(pathview)
library(gage)
library(gageData)


```

```{r}

# pull out the datasets we are interested in 

data(kegg.sets.hs)

data(sigmet.idx.hs)

```


```{r}

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)


```

The main gage() function requires a named vector of fold changes, where the names of the values are the Entrez gene IDs.

Note that we used the mapIDs() function above to obtain Entrez gene IDs (stored in res$entrez) and we have the fold change results from DESeq2 analysis (stored in res$log2FoldChange).


```{r}

foldchanges = res$log2FoldChange

names(foldchanges) = res$entrez

head(foldchanges)

```

Run the gage function on our foldchanges (with entrez annotations)
```{r}

keggres = gage(foldchanges, gsets=kegg.sets.hs)

```


Look at our results 
```{r}

attributes(keggres)

```

We want to look at the first pathways that are downregulated 
```{r}

head(keggres$less)

```

Use the pathview() function to make a pathway plot. 

First manually supply a pathway.id (namely the first part of the "hsa04110 Cell cycle") that we could see in the cell above. 


```{r}

pathview(gene.data=foldchanges, pathway.id="hsa04110")

```


```{r}

# A different PDF based output of the same data
pathview(gene.data=foldchanges, pathway.id="hsa04110", kegg.native=FALSE)

```


First we are going to pull out the top 4 pathways to then feed into pathview
```{r}

keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start=1, stop=8)
keggresids

```


Pass these IDs in keggresids to the pathview() function to draw plots for all the top 5 pathways.

```{r}

pathview(gene.data=foldchanges, pathway.id=keggresids, species="hsa")

```


# Gene Ontology

We can also do a similar procedure with gene ontology. 

```{r}

data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets=gobpsets, same.dir=TRUE)

lapply(gobpres, head)

```


# Reactome Analysis

Reactome is a database consisting of biological molecules and their relation to pathways and processes. Reactome has an R package available (https://bioconductor.org/packages/release/bioc/html/ReactomePA.html).

Using R, output the list of significant genes from our res file from DESeq2 at the 0.05 level as a plain text file:

```{r}

sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))

```

```{r}

write.table(sig_genes, file="significant_genes.txt", row.names=FALSE, col.names=FALSE, quote=FALSE)

```

Then, to perform pathway analysis online go to the Reactome website (https://reactome.org/PathwayBrowser/#TOOL=AT). Select “choose file” to upload your significant gene list. Then, select the parameters “Project to Humans”, then click “Analyze”.

Q: What pathway has the most significant “Entities p-value”? 

Endosomal/vacuolar pathway 


```{r}

sessionInfo()

```

