---
title: "Untitled"
output: github_document
---

```{r }
library(GenomicDataCommons)
library(TCGAbiolinks)
library(maftools)
```



```{r eval = FALSE}

BiocManager::install("GenomicDataCommons")

```

```{r}

status()

```
3. Querying the GDC from R
We will typically start our interaction with the GDC by searching the resource to find data that we are interested in investigating further. In GDC speak this is called “Querying GDC metadata”. Metadata here refers to the extra descriptive information associated with the actual patient data (i.e. ‘cases’) in the GDC.

For example: Our query might be ‘find how many patients were studied for each major project’ or ‘find and download all gene expression quantification data files for all pancreatic cancer patients’. We will answer both of these questions below.

The are four main sets of metadata that we can query, namely projects(), cases(), files(), and  annotations(). We will start with projects()


```{r }

projects <- getGDCprojects()
head(projects)


```

Moving onto cases() we can use an example from the package associated publication to answer our first from question above (i.e. find the number of cases/patients across different projects within the GDC):
```{r}
cases_by_project <- cases() %>%
  facet("project.project_id") %>%
  aggregations()
head(cases_by_project)
```

Q9. Write the R code to make a barplot of the cases per project. Lets plot this data with a log scale for the y axis (log="y"), rotated axis labels (las=2) and color the bar coresponding to the TCGA-PAAD project.

```{r}

x <- cases_by_project$project.project_id

# Make a custom color vector for our plot
colvec <- rep("lightblue", nrow(x))
colvec[x$key == "TCGA-PAAD"] <- "red"

# Plot with 'log' for y axis and rotate labels with 'las'
#par(___)  
barplot(x$doc_count, names.arg=x$key, log="y", col=colvec, las=2)

```

We can use the getSampleFilesSummary() function to determine for a given project how many cases and what type of data we have available for each case:

```{r}
samp <- getSampleFilesSummary("TCGA-PAAD")

```

```{r}
head(samp)

```

Now we can use GDCquery() function to focus in on a particular data type that we are interested in. For example, to answer our second question from above - namely ‘find all gene expression data files for all pancreatic cancer patients’:

```{r}
query <- GDCquery(project="TCGA-PAAD",
                  data.category="Transcriptome Profiling",
                  data.type="Gene Expression Quantification")

ans <- getResults(query)
```

```{r}

head(ans)

```

In RStudio we can now use the View() function to get a feel for the data organization and values in the returned  ans object.

```{r}


```











