---
title: "Class 13: Genome Informatics"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r }

# read in sample genotypes 
info <- read.csv("373531-SampleGenotypes-Homo_sapiens_Variation_Sample_rs8067378 (1).csv")

table(info$Genotype..forward.strand.)


```


What percentage of the population has each of the genotypes?
```{r}

(table(info$Genotype..forward.strand.)/nrow(info) *100)


```

Read this file into R and determine the sample size for each genotype and their
corresponding median expression levels for each of these genotypes.

```{r}

genotypes = read.csv("rs8067378_ENSG00000172057.6.txt", sep = "")

summary(genotypes)

```

Create a boxplot comparing genotype to the gene expression
```{r}

gg <- genotypes$exp[genotypes$geno == "G/G"]

mean(gg)

aa <- genotypes$exp[genotypes$geno == "A/A"]

mean(aa)

ag <- genotypes$exp[genotypes$geno == "A/G"]

mean(ag)

boxplot(aa, ag, gg, notch = TRUE)

```

